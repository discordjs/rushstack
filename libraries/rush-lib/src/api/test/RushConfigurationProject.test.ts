// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { validateRelativePathField } from '../RushConfigurationProject';

describe(validateRelativePathField.name, () => {
  it('accepts valid paths', () => {
    validateRelativePathField('path/to/project', 'projectFolder', '/rush.json');
    validateRelativePathField('project', 'projectFolder', '/rush.json');
    validateRelativePathField('.', 'projectFolder', '/rush.json');
    validateRelativePathField('..', 'projectFolder', '/rush.json');
    validateRelativePathField('../path/to/project', 'projectFolder', '/rush.json');
  });

  it('should throw an error if the path is not relative', () => {
    expect(() =>
      validateRelativePathField('C:/path/to/project', 'projectFolder', '/rush.json')
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      validateRelativePathField('/path/to/project', 'publishFolder', '/rush.json')
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw an error if the path ends in a trailing slash', () => {
    expect(() =>
      validateRelativePathField('path/to/project/', 'someField', '/repo/rush.json')
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      validateRelativePathField('p/', 'someField', '/repo/rush.json')
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw an error if the path contains backslashes', () => {
    expect(() =>
      validateRelativePathField('path\\to\\project', 'someField', '/repo/rush.json')
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      validateRelativePathField('path\\', 'someOtherField', '/repo/rush.json')
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw an error if the path is not normalized', () => {
    expect(() =>
      validateRelativePathField('path/../to/project', 'someField', '/repo/rush.json')
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      validateRelativePathField('path/./to/project', 'someField', '/repo/rush.json')
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      validateRelativePathField('./path/to/project', 'someField', '/repo/rush.json')
    ).toThrowErrorMatchingSnapshot();
  });
});
