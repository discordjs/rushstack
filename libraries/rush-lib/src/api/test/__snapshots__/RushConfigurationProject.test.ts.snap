// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateRelativePathField should throw an error if the path contains backslashes 1`] = `"The value \\"path\\\\to\\\\project\\" in the \\"someField\\" field in \\"/repo/rush.json\\" may not contain backslashes ('\\\\'), since they are interpreted differently on POSIX and Windows. Paths must use '/' as the path separator."`;

exports[`validateRelativePathField should throw an error if the path contains backslashes 2`] = `"The value \\"path\\\\\\" in the \\"someOtherField\\" field in \\"/repo/rush.json\\" may not contain backslashes ('\\\\'), since they are interpreted differently on POSIX and Windows. Paths must use '/' as the path separator."`;

exports[`validateRelativePathField should throw an error if the path ends in a trailing slash 1`] = `"The value \\"path/to/project/\\" in the \\"someField\\" field in \\"/repo/rush.json\\" may not end with a trailing '/' character."`;

exports[`validateRelativePathField should throw an error if the path ends in a trailing slash 2`] = `"The value \\"p/\\" in the \\"someField\\" field in \\"/repo/rush.json\\" may not end with a trailing '/' character."`;

exports[`validateRelativePathField should throw an error if the path is not normalized 1`] = `"The value \\"path/../to/project\\" in the \\"someField\\" field in \\"/repo/rush.json\\" should be replaced with its normalized form \\"to/project\\"."`;

exports[`validateRelativePathField should throw an error if the path is not normalized 2`] = `"The value \\"path/./to/project\\" in the \\"someField\\" field in \\"/repo/rush.json\\" should be replaced with its normalized form \\"path/to/project\\"."`;

exports[`validateRelativePathField should throw an error if the path is not normalized 3`] = `"The value \\"./path/to/project\\" in the \\"someField\\" field in \\"/repo/rush.json\\" should be replaced with its normalized form \\"path/to/project\\"."`;

exports[`validateRelativePathField should throw an error if the path is not relative 1`] = `"The value \\"C:/path/to/project\\" in the \\"projectFolder\\" field in \\"/rush.json\\" must be a relative path."`;

exports[`validateRelativePathField should throw an error if the path is not relative 2`] = `"The value \\"/path/to/project\\" in the \\"publishFolder\\" field in \\"/rush.json\\" must be a relative path."`;
